File: runme/mcgen/local/datan/exasrc/e2ds.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__5 = 5, c__27 = 27;
static double c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, j, k, n, nint, nf, idelnf, nx, nf0, nws;
    static double delf, delx, hist[1000], f[1000], r, f0, x0;
    static char capt[28], tx[5], ty[5];
    /* System generated locals */
    integer i__1, i__2;
/* identify program to user */
    printf("%s\n", "Program E2DS simulates experiment");
    printf("%s\n", "of Rutherford and Geiger");
    printf("%s\n\n", "on statistical nature of radioactive decays.");
/* ask for numerical input */
    printf("%s\n", "Enter number N of observed decays:");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter number NINT of time intervals (1 <= NINT <= 1000):");
    printf("%s", "> ");
    scanf("%li", &nint);
/* prepare histogram */
    x0 = 0.;
    delx = 1.;
    nx = nint;
    nf = (integer) ((double) n / (double) nint);
    idelnf = (integer) sqrt((double) nf) + 1;
/* Computing MAX */
    i__1 = 0, i__2 = nf - idelnf * 5;
    nf0 = max(i__1,i__2);
    nf += idelnf * 5;
    f0 = -.5;
    delf = 1.;
    nf = min(nf, 1000);
/* prepare texts */
    sprintf(capt, "N = %6li, n_int# = %6li", n, nint);
    smhsin_(hist, &x0, &delx, &nx);
/* simulate decay times and fill histogram */
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
	rnecuy_(&r, &c__1);
	r *= (double) nint;
	smhsfl_(hist, &x0, &delx, &nx, &r, &c_d1);
    }
/* graphical output */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* present histogram */
    memset(tx, ' ', 5);
    memset(ty, ' ', 5);
    strncpy(tx, "t", 1);
    strncpy(ty, "N(t)", 4);
    smhsgr_(hist, &x0, &delx, &nx, tx, &c__5, ty, &c__5, capt, &c__27, &nws);
/* analyze histogram of decay times */
    i__1 = nf + 1;
    for (i = 1; i <= i__1; ++i) {
	k = i - 1;
	f[i - 1] = 0.;
	i__2 = nx;
	for (j = 1; j <= i__2; ++j) {
	    if ((integer) hist[j - 1] == k) {
		f[i - 1] += 1.;
	    }
	}
    }
    strncpy(tx, "k", 1);
    strncpy(ty, "N(k)", 4);
    smhsgr_(f, &f0, &delf, &nf, tx, &c__5, ty, &c__5, capt, &c__27, &nws);
    return 0;
} /* main */

File: runme/mcgen/local/datan/exasrc/e2ds.f
      PROGRAM E2DS
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION HIST(1000),F(1000)
      CHARACTER*5 TX,TY
      CHARACTER*27 CAPT
C identify program to user
      WRITE(*,*)' Program E2DS simulates experiment'
      WRITE(*,*)' of Rutherford and Geiger'
      WRITE(*,*)' on statistical nature of radioactive decays'
      WRITE(*,*)' '
C ask for numerical input
      WRITE(*,*)' Enter number N of observed decays'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)
     +' Enter number NINT of time intervals (1 <= NINT <= 1000)'
      WRITE(*,*)' >'
      READ(*,*)NINT
C prepare histogram
      X0=0.D0
      DELX=1.D0
      NX=NINT
      NF=INT(FLOAT(N)/FLOAT(NINT))
      IDELNF=INT(SQRT(FLOAT(NF)))+1
      NF0=MAX(0,NF-5*IDELNF)
      NF=NF+5*IDELNF
      F0=-0.5D0
      DELF=1.D0
      NF=MIN(NF,1000)
C prepare texts
      CAPT='N = ******, n_int# = ******'
      WRITE(CAPT(5:10),'(I6)')N
      WRITE(CAPT(22:27),'(I6)')NINT
      CALL SMHSIN(HIST,X0,DELX,NX)
C simulate decay times and fill histogram
      DO 10 I=1,N
        CALL RNECUY(R,1)
        R=R*DBLE(NINT)
        CALL SMHSFL(HIST,X0,DELX,NX,R,1.D0)
10    CONTINUE
C graphical output
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C present histogram
      TX='t'
      TY='N(t)'
      CALL SMHSGR(HIST,X0,DELX,NX,TX,5,TY,5,CAPT,27,NWS)
C analyze histogram of decay times
      DO 30 I=1,NF+1
        K=I-1
        F(I)=0.D0
        DO 20 J=1,NX
          IF(INT(HIST(J)).EQ.K) F(I)=F(I)+1.D0
20      CONTINUE
30    CONTINUE
      TX='k'
      TY='N(k)'
      CALL SMHSGR(F,F0,DELF,NF,TX,5,TY,5,CAPT,27,NWS)
      END
